ext/s3/writer.go

```diff
@@ -3,106 +3,132 @@ package s3
 import (
 	"context"
 	"io"
-	"sync"
+	"os"
 
-	"github.com/aws/aws-sdk-go-v2/aws"
-	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
 	"github.com/aws/aws-sdk-go-v2/service/s3"
-	"github.com/djherbis/buffer"
-	"github.com/djherbis/nio/v3"
 	"github.com/pkg/errors"
 )
 
 // S3Writer implements io.Writer and io.ReaderFrom for streaming data to S3
 type S3Writer struct {
 	ctx        context.Context
-	uploader   *manager.Uploader
+	client     *s3.Client
 	bucketName string
 	key        string
-	pipeReader *nio.PipeReader
-	pipeWriter *nio.PipeWriter
-	errChan    chan error
-	closed     bool
-	mu         sync.Mutex
+	// internal use
+	chunkSize int      // size of the chunk to write to S3, default is 64MB
+	fileTmp   *os.File // temporary file to write the data until flush
+	size      int      // size of the data written to the temporary file
 }
 
 // NewS3Writer creates a new writer that streams data to S3
-func NewS3Writer(ctx context.Context, uploader *manager.Uploader, bucketName, key string) *S3Writer {
-	buf := buffer.New(32 * 1024)
-	pr, pw := nio.Pipe(buf)
-	errChan := make(chan error, 1)
-
+func NewS3Writer(ctx context.Context, client *s3.Client, bucketName, key string) (*S3Writer, error) {
 	w := &S3Writer{
 		ctx:        ctx,
-		uploader:   uploader,
+		client:     client,
 		bucketName: bucketName,
 		key:        key,
-		pipeReader: pr,
-		pipeWriter: pw,
-		errChan:    errChan,
+		chunkSize:  (1 << 20) * 64, // default chunk size is 64MB
+		size:       0,
```

Comment:
are we duplicating the chunkWriter mechanism here?

Reply:
yes mas, to avoid multiple api call upon `Write` and reduce memory footprint